operator/pkg/ciliumidentity/controller.go

```diff
@@ -108,15 +132,68 @@ func (c *Controller) Start(_ cell.HookContext) error {
 		return fmt.Errorf("cid controller failed to calculate the desired state: %w", err)
 	}
 
+	// The Cilium Identity (CID) controller running in cilium-operator is
+	// responsible only for managing CID API objects.
+	//
+	// Pod events are added to Pod work queue.
+	// Namespace events are processed immediately and added to Pod work queue.
+	// CID events are added to CID work queue.
+	// Processing Pod work queue items are adding items to CID work queue.
+	// Processed CID work queue items result in mutations to CID API objects.
+	//
+	// Diagram:
+	//-------------------------Pod event---------CID event
+	//---------------------------||-----------------||
+	//----------------------------V------------------V
+	// Namespace event -> Pod work queue -> CID work queue -> Mutate CID API objects
+	c.wp = workerpool.New(workerPoolSize)
+	err = c.startEventProcessing()
+	if err != nil {
+		return err
+	}
+	err = c.startWorkQueues()
+	if err != nil {
+		return err
+	}
+
 	return nil
 }
 
-func (c *Controller) enqueueCIDReconciliation(cidKey resource.Key, delay time.Duration) {
+func (c *Controller) Stop(_ cell.HookContext) error {
+	c.cidQueue.ShutDown()
+	c.podQueue.ShutDown()
+
+	c.contextCancel()
+	return c.wp.Close()
 }
 
-func (c *Controller) enqueuePodReconciliation(podKey resource.Key, delay time.Duration) {
+func (c *Controller) initializeQueues() {
+	c.initCIDQueue()
+	c.initPodQueue()
 }
 
-func (c *Controller) Stop(hookContext cell.HookContext) error {
+func (c *Controller) startEventProcessing() error {
+	if err := c.wp.Submit("proc-cid-events", c.processCiliumIdentityEvents); err != nil {
+		return fmt.Errorf("failed to init worker pool for CiliumIdentityEvents")
```

Comment:
Do you want to chain the original error so it will be logged?

`fmt.Errorf("...: %w", err)`

If you don't have formatting done, recommend using `errors.New()` instead.